### DISCUSSION TITLE
Roadmap 2022/23

### DISCUSSION THREAD

**Discussion by joshuaellis** [pmndrs/react-spring#1980] [Announcements]
# The `react-spring` roadmap 2022 / 23

I believe this is the first roadmap that `react-spring` has seriously had documented beside holding RFCs for breaking change versions. Its looking at the next 12-18 months from now till the end of 2023, so the next review of the roadmap will most likely at earliest be Aug/Sept 2023 and will be based off this one, potentially carrying areas that were not tackled but are still areas of interest.

This roadmap is not definitive & can be modified to "remove" items (nothing will be removed, more just striken through) as well as refining particular areas of the roadmap. Some initiatives will be described as "investigation" because without real understanding of the problem, committing to actual feature work is not realistic.

As always, feedback is well received and contributions are more than welcome either through actual work or financial to aid the development of the library – https://opencollective.com/react-spring

---

### Legend

| Mark | Description |
| ------ | ---------- |
| Empty Check | Work not started |
| Checked | Work finished |
| ✍️ | Work in progress |
| 🛑 | Blocked| 

## Documentation

- [x] Finish moving / re-writing documentation to the new beta-docs site
- [x] Design & build examples directory for doc site
- [ ] ✍️ Add more generic / basic use case examples
	- [ ] Slide-in side panel
	- [x] CSS Gradients
	- [ ] Simple modal
	- [ ] Slide-up drawer (think native styled)
	- [ ] Compass (shows full 360º rotation)
- [ ] Design & build tutorials directory
- [ ] Create 3 tutorials

## Performance

- [ ] Investigate how we can support `sideEffects: false` again
- [ ] Investigate react-native performance, how does it compare to re-animated?
- [ ] Focus further on bundle size, where can we trim the fat?

## Features & Fixes

- [ ] ✍️ Investigate how we can best support different `raf` drivers in one environment (e.g. r3f & web vs globals package)
- [ ] Animate Presence & Animated layout components
- [ ] Timelines, can we support them and how would this look / work?
- [ ] Introduce more utility hooks, perhaps as an off-shoot library that includes things like "onHover" etc.
	- [x] useScroll
	- [x] useInView
	- [ ] useAnimationFrame
	- [ ] useDrag(?)

## General

- [ ] Look at deprecating the use of `react-spring` npm library
	- Requires new docs to go live that sets the best practice of scoped packages
- [ ] Continue to make onboarding for new developers easier and more attractive
- [ ] Explore how to engage with our community further increasing contributions to the project

---

## Ending Notes

The key takeaway for the past year has been that our documentation / examples etc. are all geared towards power users, they can be complicated and hard to follow and not enough of the simple stuff is documented, off putting new devs to try the library, hopefully we can begin to fix this whilst not forgetting about our "power users".

Thanks for reading this and always, if you have any thoughts and/or opinions then please, feel free to share them here as part of healthy discussion!

**Comment by leo-petrucci** [comment]
Some really exciting stuff ahead 👀

I'm not great at maintaining libraries but I'm slightly better at building examples. Is there a list of things you'd like to have as examples for the docs? I could probably take a bash at it whenever I have time.

**Reply by joshuaellis** [reply]
I'll write a couple up under the bullet point, to give you an _idea_ of what i'm thinking, but I'm happy for this to be community driven so things like your infinite slider would work too!

**Reply by leo-petrucci** [reply]
The code for that could use some improvement but I can probably give it a slight refresh 👍
I'd love for react spring to have more "real world" examples, like how to integrate it with radix-ui components for example.

**Reply by joshuaellis** [reply]
https://beta.react-spring.dev/docs/concepts/animated-elements#example-using-stitches--radix-ui definitely, they're important to show 👀

**Comment by heyfirst** [comment]
I'm really exciting on this roadmap! I'm a big fan of this library and the original docs is already really awesome anyway. Well, also agree with you that examples in original docs is too advance to understand a basic. Having more basic example will be lower the difficulty for new entry for sure (and IMHO, I'm sure they will be more easy to have fun with animation and this library 😄)

So, Does anyway I can help on writing beta docs?

I feel like I can help to grasp some useful thing from the original docs to new one. for example the spring preset on [this beta docs](https://beta.react-spring.dev/docs/advanced/config#presets). I like the way the original docs present which is the config with codesandbox demo and then the preset with codesandbox (https://react-spring.dev/common/configs#configs).

This is my first-experience on this config document.
- the first one, the config helps me to understand how `mass`, `tension`, `friction` affect the animation.
- the second one, the config preset show me the example how to use the `mass`, `tension`, `friction` properly,
- then I go back to adjust the `mass`, `tension`, `friction` by myself once again to find others movement I love 🥳 

IMHO, it's such a great document page!

Also, if you have some your document backlog to move from the original to the beta one. I can pick some and help you!

**Reply by heyfirst** [reply]
I guess it is this one (https://github.com/pmndrs/react-spring/projects/3) 🙇🏻

**Reply by joshuaellis** [reply]
hey @heyfirst,

Thanks for the positive feedback. Great to hear from the community.

> So, Does anyway I can help on writing beta docs?

Absolutely, you're totally welcome to contribute things such as the config visualiser you're talking about. You can probably submit a PR directly and we can discuss it further there.

> Also, if you have some your document backlog to move from the original to the beta one. I can pick some and help you!

Yep, it's [pmndrs/react-spring/projects/3](https://github.com/pmndrs/react-spring/projects/3) in general, I might have missed things however, so you could also open issues to discuss these matters prior to you doing the work as that might be more appropriate.

**Reply by heyfirst** [reply]
> Absolutely, you're totally welcome to contribute things such as the config visualiser you're talking about. You can probably submit a PR directly and we can discuss it further there.

Sure @joshuaellis! I will open some PR for this small doc update by tomorrow or the day after! and sure I will open some issues to discuss about beta docs if I see some can be added 🥳
